packages/@lwc/engine-dom/src/custom-elements/create-custom-element.ts

```diff
@@ -92,12 +91,18 @@ export function getUpgradableConstructor(tagName: string) {
 export const createCustomElement = (
     tagName: string,
     upgradeCallback: LifecycleCallback,
-    useNativeLifecycle: boolean
+    useNativeLifecycle: boolean,
+    isFormAssociated: boolean | undefined
 ) => {
-    const UpgradableConstructor = getUpgradableConstructor(tagName);
+    const UpgradableConstructor = getUpgradableConstructor(tagName, isFormAssociated);
 
     elementBeingUpgradedByLWC = true;
     try {
+        if (UpgradableConstructor.formAssociated !== isFormAssociated) {
+            throw new Error(
+                `<${tagName}> was already registered with formAssociated=${UpgradableConstructor.formAssociated}. It cannot be re-registered with formAssociated=${isFormAssociated}. Please rename your component to have a different name than <${tagName}>`
```

Comment:
The original issue called out adding a link to the lwc.dev guide but the info on form-associated custom elements and `ElementInternals` isn't available yet.

Reply:
I'm intentionally throwing the error here rather than in the `base-lightning-element` because I wanted to avoid creating a new instance if we already know it should throw an error.